app-modules/prospect/src/Imports/ProspectImporter.php

```diff
@@ -0,0 +1,154 @@
+
+                    if ($source) {
+                        return;
+                    }
+
+                    $fail('The selected source is invalid.');
+                }])
+                ->guess(['source'])
+                ->fillRecordUsing(function (Prospect $record, string $state) {
+                    $source = ProspectSource::query()
+                        ->where('name', $state)
+                        ->first();
+
+                    if (! $source) {
+                        return;
+                    }
+
+                    $record->source()->associate($source);
+                })
+                ->requiredMapping(),
+            ImportColumn::make('description'),
+            ImportColumn::make('email')
+                ->rules(['email']),
+            ImportColumn::make('email_2')
+                ->rules(['email']),
+            ImportColumn::make('mobile'),
+            ImportColumn::make('sms_opt_out')
+                ->boolean()
+                ->rules(['boolean']),
+            ImportColumn::make('email_bounce')
+                ->boolean()
+                ->rules(['boolean']),
+            ImportColumn::make('phone'),
+            ImportColumn::make('address'),
+            ImportColumn::make('address_2'),
+            ImportColumn::make('birthdate')
+                ->rules(['date']),
+            ImportColumn::make('hsgrad'),
+        ];
+    }
+
+    public function resolveRecord(): ?Model
+    {
+        $email = $this->data['email'] ?? null;
+        $email2 = $this->data['email_2'] ?? null;
+
+        $emails = array_filter(
+            [$email, $email2],
+            fn (mixed $state): bool => filled($state),
+        );
+
+        if (empty($emails)) {
+            return new Prospect();
+        }
+
+        $prospect = Prospect::query()
+            ->whereIn('email', $emails)
+            ->orWhereIn('email_2', $emails)
+            ->first();
+
+        if ($prospect) {
+            return $prospect;
+        }
+
+        return new Prospect([
+            'email' => $email,
+            'email_2' => $email2,
+        ]);
+    }
+
+    public function beforeCreate(): void
+    {
+        $this->record->assignedTo()->associate($this->import->user);
+        $this->record->createdBy()->associate($this->import->user);
+    }
+
+    public static function getCompletedNotificationBody(int $totalRows): string
+    {
+        return "Your prospect import has completed and {$totalRows} rows were processed.";
+    }
+
+    public static function getFailureNotificationBody(int $processedRows): string
+    {
+        return "Something went wrong after importing {$processedRows} rows of prospects.";
+    }
+
+    public function getValidationFailureNotificationBody(string $message): string
+    {
+        return "A validation error occurred while importing a prospect: {$message}. This skipped row belonged to the prospect with full name \"{$this->data['full_name']}\".";
+    }
```

Comment:
Does this mean that we import rows individually? So if one row fails, that doesn't prevent the other rows from importing?

Reply:
Correct, and we send a notification for each row that fails validation, with the specific validation error. As it can differ.